contribulyze/contribulyze.py

```diff
@@ -0,0 +1,663 @@
+  def __str__(self):
+    s = 'FIELD: %s (%d contributors)\n' % (self.name, len(self.contributors))
+    for contributor in self.contributors:
+      s += str(contributor) + '\n'
+    s += self.addendum
+    return s
+
+
+class LogMessage(object):
+  # Maps sha strings onto LogMessage instances,
+  # holding all the LogMessage instances ever created.
+  all_logs = { }
+  author = None
+  date = None
+  message = ''
+  latest = None
+  def __init__(self, sha):
+    """Instantiate a log message.  All arguments are strings, including commit."""
+    self.sha = sha
+    # Map field names (e.g., "Patch", "Review") onto Field objects.
+    self.fields = { }
+    if sha in LogMessage.all_logs:
+      complain("SHA '%s' seen more than once" % sha, True)
+    LogMessage.all_logs[sha] = self
+  def add_field(self, field):
+    self.fields[field.name] = field
+  def accum(self, line):
+    """Accumulate one more line of raw message."""
+    self.message += line
+
+  def __cmp__(self, other):
+    """Compare two log messages by date, for sort().
+    Return -1, 0 or 1 depending on whether a > b, a == b, or a < b.
+    Note that this is reversed from normal sorting behavior, but it's
+    what we want for reverse chronological ordering of revisions."""
+    a = self.date
+    b = other.date
+    if a > b: return -1
+    if a < b: return 1
+    else:     return 0
+
+  def sort_key(self):
+    return self.date
+
+  def __str__(self):
+    s = '=' * 15
+    header = ' COMMIT: %s | %s \n %s' % (self.sha, self.author, self.message)
+    s += header
+    s += '=' * 15
+    s += '\n'
+    for field_name in self.fields.keys():
+      s += str(self.fields[field_name]) + '\n'
+    s += '-' * 15
+    s += '-' * len(header)
+    s += '-' * 15
+    s += '\n'
+    return s
+
+def process_aliases(aliases_input):
+  line = aliases_input.readline()
+  while line:
+    aliases = line.split(',')
+    c = Contributor.get(aliases.pop(0).strip(), None)
+    for alias in aliases:
+      c.add_aliases(alias.strip())
+    line = aliases_input.readline()
+
+def process_committers():
+    committers_url = 'https://people.apache.org/public/public_ldap_groups.json'
+    names_url = 'https://people.apache.org/public/icla-info.json'
+    committers = json.loads(requests.get(committers_url).text)['groups']['cassandra']['roster']
+    names_json = json.loads(requests.get(names_url).text)['committers']
+    for committer in committers:
+        name = names_json[committer]
+        if committer in Contributor.all_contributors and not name in Contributor.all_contributors:
+            c = Contributor.get(committer, None)
+            c.add_aliases(name)
+        else:
+            c = Contributor.get(name, None)
+            c.add_aliases(committer)
+        c.is_committer = True
+
+
+### Regexps to parse the logs. ##
+log_header_re = re.compile('^commit ([0-9a-z]+)$')
+patch_by_re = re.compile('(?:.*\n)*.*patch by ([^;]+)(;|,)', flags=re.IGNORECASE | re.MULTILINE)
```

Comment:
@michaelsembwever 
I feel there is an issue here. Reviewers are parsed as contributors when there is a comma at the end of contributor names and a comma in the list of reviewers, for example: "patch by Berenguer Blasi, reviewed by Tomasz Lasica, Andres de la Peña and Ekaterina Dimitrova". Basically everything until the last comma is considered to be a contributors list. As a result the script creates wrong records in the output: https://nightlies.apache.org/cassandra/devbranch/misc/contribulyze/html/all_time/detail/reviewed%2520by%2520Tomasz%2520Lasica.html (there are more like this).

I think we need to stop when we see start of the reviewers section. Smth like this:
```
patch_by_re = re.compile('(?:.*\n)*.*patch by ([^;]+)[;, ](?:review|test)(?:ed)? by', flags=re.IGNORECASE | re.MULTILINE)
```

I briefly tested it on a few examples and it seems to be working fine, but I had not put a lot of effort into the testing process.

Reply:
thanks @alex-ninja , let me try it out in an extra commit on the branch…

Reply:
[that](https://github.com/apache/cassandra-builds/pull/54/commits/ab03d1758a9d5e11aad45ba1fcc14ab09c35edb3) didn't work.

Example results:

<img width="617" alt="Screenshot 2021-09-17 at 18 44 15" src="https://user-images.githubusercontent.com/559444/133824384-a77a00fe-c59a-4312-9c5f-357dafe7de9e.png">

<img width="617" alt="Screenshot 2021-09-17 at 18 45 05" src="https://user-images.githubusercontent.com/559444/133824495-5bb72ce8-53ab-4c2d-8c82-a847a660eea1.png">

lots of missing "patches" in those lists. removing that commit for now…